src/v/coproc/reconciliation_backend.cc

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/reconciliation_backend.h"
+
+#include "cluster/cluster_utils.h"
+#include "cluster/shard_table.h"
+#include "cluster/topic_table.h"
+#include "coproc/logger.h"
+#include "coproc/pacemaker.h"
+#include "storage/api.h"
+
+#include <seastar/core/coroutine.hh>
+
+namespace coproc {
+
+reconciliation_backend::reconciliation_backend(
+  ss::sharded<cluster::topic_table>& topics,
+  ss::sharded<cluster::shard_table>& shard_table,
+  ss::sharded<storage::api>& storage) noexcept
+  : _self(model::node_id(config::node().node_id))
+  , _data_directory(config::node().data_directory().as_sstring())
+  , _topics(topics)
+  , _shard_table(shard_table)
+  , _storage(storage) {
+    _retry_timer.set_callback([this] {
+        (void)within_context([this]() { return fetch_and_reconcile(); });
+    });
+}
+
+template<typename Fn>
+ss::future<> reconciliation_backend::within_context(Fn&& fn) {
+    return ss::with_gate(_gate, [this, fn = std::forward<Fn>(fn)]() mutable {
+        if (!_as.abort_requested()) {
+            return _mutex.with(
+              [fn = std::forward<Fn>(fn)]() mutable { return fn(); });
+        }
+        return ss::now();
+    });
+}
+
+ss::future<> reconciliation_backend::start() {
+    _id_cb = _topics.local().register_delta_notification(
```

Comment:
are we sure that we are not missing any of the notifications ?

Reply:
What do you mean? Like am I sure that the callbacks are fired when expected? Well there was that PR a while back I got merged for that behavior that was broken. These tests were failing due to that, now they pass.

Reply:
maybe Michal is asking if there are notifications missing because registration happens after an event is processed from the controller log. I'm not sure though if that is the actual question.

Reply:
In either case this is something I did not think about, i will investigate

Reply:
Yes so I have modified the lifetime of the `coproc::partition_manager` to be constructed and destructed immediately after & before the `cluster::partition_manager` so I am assuming these issues can no longer occur even though I didn't attempt to explicitly test this corner case.